### PR TITLE
Add support for single-digits in cudf::to_timestamps

### LINKED ISSUE #9082: [BUG] `to_datetime` does not correctly convert single digit  months and hours to datetime
Non zero padded months and hours should be correctly parsed 

image

but changing string to `'2010-08-21 08:43'` will parse correctly

**Expected behavior**
parse the non-zero padded months and hours (and possibly others) correctly

Here's a workaround example for month:

```
month = df.month.astype(str)
month.loc[month.str.len() == 1] = '0' + month.loc[month.str.len() == 1]
```
Rapids 21.06 - Docker install

**Additional context**
Used for parsing dates that are later sorted - so this issue is very important for temporal work.

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #9082 

This PR adds support to the `cudf::to_timestamps` API to allow single digit day and month in the input strings format.

Corresponding updates are also made to `cudf::is_timestamps`.

### CODE REVIEW

File: cpp/src/strings/convert/convert_datetime.cu [hunk 1]
```diff
@@ -416,18 +439,23 @@ struct check_datetime_format {
    * @param bytes Number of bytes to check.
    * @param min_value Inclusive minimum value
    * @param max_value Inclusive maximum value
-   * @return true if parsed value is between `min_value` and `max_value`.
+   * @return number of bytes not successfully processed
    */
-  __device__ bool check_value(const char* str, size_type bytes, int min_value, int max_value)
+  __device__ size_type check_value(const char* str,
```
Comment (devavret) [inline cpp/src/strings/convert/convert_datetime.cu:444]: Feels smelly to return only bytes for a function that's supposed to do a checking. Consider returning pair<bool, size_type>.

**Review by devavret — COMMENTED** [review]
LGTM otherwise.

File: cpp/src/strings/convert/convert_datetime.cu [hunk 2]
```diff
@@ -159,6 +159,25 @@ struct format_compiler {
...
 
+/**
+ * @brief Specialized function to return the integer value reading up to the specified
+ * bytes or until an invalid character is encountered.
+ *
+ * @param str Beginning of characters to read.
+ * @param bytes Number of bytes in str to read.
+ * @return Integer value of valid characters read and how many bytes were not read.
+ */
+__device__ thrust::pair<int32_t, size_type> parse_int(const char* str, size_type bytes)
```
Comment (vyasr) [inline cpp/src/strings/convert/convert_datetime.cu:170]:
```suggestion
__device__ thrust::pair<int32_t, size_type> parse_int(char const* str, size_type bytes)
```

Comment (vyasr) [reply cpp/src/strings/convert/convert_datetime.cu:170]: This is the rare instance where (maybe just due to being accustomed to the pattern) it seems like passing bytes by reference and modifying would be cleaner than returning a pair that needs to be unpacked later, but I guess that goes against the `libcudf` style guide?

File: cpp/src/strings/convert/convert_datetime.cu [hunk 3]
```diff
@@ -168,35 +187,22 @@ template <typename T>
...
-  int8_t subsecond_precision;
+  int8_t const subsecond_precision;
 
   /**
    * @brief Return power of ten value given an exponent.
    *
    * @return `1x10^exponent` for `0 <= exponent <= 9`
    */
-  __device__ constexpr int64_t power_of_ten(int32_t exponent)
+  __device__ constexpr int64_t power_of_ten(int32_t exponent) const
```
Comment (vyasr) [inline cpp/src/strings/convert/convert_datetime.cu:197]:
```suggestion
  __device__ constexpr int64_t power_of_ten(int32_t const exponent) const
```

File: cpp/src/strings/convert/convert_datetime.cu [hunk 4]
```diff
@@ -208,32 +214,71 @@ struct parse_datetime {
...
+          auto const [year, left] = parse_int(ptr, item.length);
+          timeparts.year          = static_cast<int16_t>(year);
+          bytes_read -= left;
+          break;
+        }
         case 'y': {
-          auto const year = str2int(ptr, item.length);
-          timeparts.year  = static_cast<int16_t>(year + (year < 69 ? 2000 : 1900));
+          auto const [year, left] = parse_int(ptr, item.length);
+          timeparts.year          = static_cast<int16_t>(year + (year < 69 ? 2000 : 1900));
```
Comment (vyasr) [inline cpp/src/strings/convert/convert_datetime.cu:234]: But what happens in 2069 😮

**Review by vyasr — COMMENTED** [review]
Some minor suggestions, otherwise LGTM.

File: cpp/src/strings/convert/convert_datetime.cu (hunk 4 above)
Comment (devavret) [reply cpp/src/strings/convert/convert_datetime.cu:234]: [This](https://nvidia.slack.com/archives/CDTANRCTT/p1609862889009600?thread_ts=1609862482.009400&cid=CDTANRCTT) and then [this](https://nvidia.slack.com/archives/CDTANRCTT/p1609878276014900?thread_ts=1609862482.009400&cid=CDTANRCTT)

File: cpp/src/strings/convert/convert_datetime.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/strings/convert/convert_datetime.cu:170]: Yes. It was actually doing this before and I switched it here (or maybe a previous PR) to better communicate what is being returned.

**Review by vyasr — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**